File: src/AoC13/Day13.java
package AoC13;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.*;

public class Day13 {

    public static void main(String[] args) throws FileNotFoundException {
        File file = new File("src/AoC13/data");
        BufferedReader br = new BufferedReader(new FileReader(file));

        List<String> lines = br.lines().toList();

        part1(lines);
        part2(lines);
    }

    public static void part2(List<String> lines) {
        List<List<String>> blocks = parseBlocks1(lines);
        long result = 0;
        for (List<String> block: blocks) {
            Optional<Integer> column = findReflectionColumnWithSmudge(block);
            Optional<Integer> row = findReflectionRowsWithSmudge(block);
            if (column.isPresent()) {
                result += column.get();
            }
            if (row.isPresent()) {
                result += 100L * row.get();
            }
        }

        System.out.println(result);
    }

    public static boolean areSmudges(String line1, String line2) {
        int differences = 0;
        for (int i = 0; i < line1.length(); ++i) {
            if (line1.charAt(i) != line2.charAt(i)) {
                ++differences;
            }
        }
        return differences == 1;
    }

    public static Optional<Integer> findReflectionRowsWithSmudge(List<String> block) {
        int m = block.size();

        for (int i = 0; i < m - 1; ++i) {
            int left = i;
            int right = i + 1;
            boolean isValid = true;
            int differences = 1;
            while (0 <= left && right < m) {
                if (!block.get(left).equals(block.get(right)) && !areSmudges(block.get(left), block.get(right))) {
                    isValid = false;
                    break;
                }
                if (areSmudges(block.get(left), block.get(right))) {
                    --differences;
                }
                if (differences < 0) {
                    isValid = false;
                    break;
                }
                --left; ++right;
            }
            if (isValid && differences == 0) {
                return Optional.of(i + 1);
            }
        }
        return Optional.empty();
    }

    public static Optional<Integer> findReflectionColumnWithSmudge(List<String> block) {
        List<String> columns = new ArrayList<>();
        int m = block.get(0).length();
        for (int i = 0; i < m; ++i) {
            StringBuilder column = new StringBuilder();
            for (String s : block) {
                column.append(s.charAt(i));
            }
            columns.add(column.toString());
        }

        for (int i = 0; i < m - 1; ++i) {
            int left = i;
            int right = i + 1;
            boolean isValid = true;
            int differences = 1;
            while (0 <= left && right < m) {
                if (!columns.get(left).equals(columns.get(right)) && !areSmudges(columns.get(left), columns.get(right))) {
                    isValid = false;
                    break;
                }
                if (areSmudges(columns.get(left), columns.get(right))) {
                    --differences;
                }
                if (differences < 0) {
                    isValid = false;
                    break;
                }
                --left; ++right;
            }
            if (isValid && differences == 0) {
                return Optional.of(i + 1);
            }
        }
        return Optional.empty();
    }

    public static void part1(List<String> lines) {
        List<List<String>> blocks = parseBlocks1(lines);
        long result = 0;
        for (List<String> block: blocks) {
            Optional<List<Integer>> columns = findReflectionColumn(block);
            Optional<List<Integer>> rows = findReflectionRows(block);
            if (columns.isPresent()) {
                for (int column : columns.get())
                    result += column;
            }
            if (rows.isPresent()) {
                for (int row : rows.get())
                    result += 100L * row;
            }
        }

        System.out.println(result);
    }

    public static Optional<List<Integer>> findReflectionRows(List<String> block) {
        int m = block.size();
        List<Integer> results = new ArrayList<>();
        for (int i = 0; i < m - 1; ++i) {
            int left = i;
            int right = i + 1;
            boolean isValid = true;
            while (0 <= left && right < m) {
                if (!block.get(left).equals(block.get(right))) {
                    isValid = false;
                    break;
                }
                --left; ++right;
            }
            if (isValid) {
                results.add(i + 1);
            }
        }
        return results.isEmpty() ? Optional.empty() : Optional.of(results);
    }

    public static Optional<List<Integer>> findReflectionColumn(List<String> block) {
        List<String> columns = new ArrayList<>();
        int m = block.get(0).length();
        for (int i = 0; i < m; ++i) {
            StringBuilder column = new StringBuilder();
            for (String s : block) {
                column.append(s.charAt(i));
            }
            columns.add(column.toString());
        }
        List<Integer> results = new ArrayList<>();
        for (int i = 0; i < m - 1; ++i) {
            int left = i;
            int right = i + 1;
            boolean isValid = true;
            while (0 <= left && right < m) {
                if (!columns.get(left).equals(columns.get(right))) {
                    isValid = false;
                    break;
                }
                --left; ++right;
            }
            if (isValid) {
                results.add(i + 1);
            }
        }
        return results.isEmpty() ? Optional.empty() : Optional.of(results);
    }

    public static List<List<String>> parseBlocks1(List<String> lines) {
        List<List<String>> result = new ArrayList<>();
        Iterator<String> iterator = lines.iterator();
        List<String> block = new ArrayList<>();
        while (iterator.hasNext()) {
            String line = iterator.next();
            if (line.isEmpty()) {
                result.add(block);
                block = new ArrayList<>();
                continue;
            }
            block.add(line);
        }
        result.add(block);
        return result;
    }

}

File: src/AoC13/Day13.py
with open(file=r'Data', mode='r') as file:
    lines = file.read().splitlines()

blocks: list[list[str]] = [[]]
for line in lines:
    if len(line) == 0:
        blocks.append([])
        continue
    blocks[-1].append(line)

def are_smudged(line1: str, line2: str) -> bool:
    return [chr1 == chr2 for chr1, chr2 in zip(line1, line2)].count(False) == 1

result: int = 0
for block in blocks:
    rows = block
    cols = ["".join(t) for t in zip(*rows)]

    refl_row: int = -1
    refl_col: int = -1

    # Check rows
    for i in range(len(rows) - 1):
        top = i
        bottom = i + 1
        is_valid = True
        diffs = 1
        while 0 <= top and bottom < len(rows):
            if rows[top] != rows[bottom] and not are_smudged(rows[top], rows[bottom]):
                is_valid = False
                break
            if are_smudged(rows[top], rows[bottom]):
                diffs -= 1
            if diffs < 0:
                is_valid = False
                break
            top -= 1
            bottom += 1
        if is_valid and diffs == 0:
            refl_row = i + 1
            break
    # Check columns
    for i in range(len(cols) - 1):
        left = i
        right = i + 1
        is_valid = True
        diffs = 1
        while 0 <= left and right < len(cols):
            if cols[left] != cols[right] and not are_smudged(cols[left], cols[right]):
                is_valid = False
                break
            if are_smudged(cols[left], cols[right]):
                diffs -= 1
            if diffs < 0:
                is_valid = False
                break
            left -= 1
            right += 1
        if is_valid and diffs == 0:
            refl_col = i + 1
            break
    if refl_row != -1:
        result += 100 * refl_row
    if refl_col != -1:
        result += refl_col

print(result)